spec/Section 6 -- Execution.md

```diff
@@ -287,17 +294,33 @@ operation type.
 
 #### Response Stream
 
-Each event in the underlying Source Stream triggers execution of the
-subscription _selection set_ using that event as a root value.
+Each event from the underlying Source Stream triggers execution of the
+subscription _selection set_ using that event's value as the {initialValue}.
 
 MapSourceToResponseEvent(sourceStream, subscription, schema, variableValues):
 
-- Return a new event stream {responseStream} which yields events as follows:
-  - For each {event} on {sourceStream}:
-    - Let {response} be the result of running
-      {ExecuteSubscriptionEvent(subscription, schema, variableValues, event)}.
-    - Yield an event containing {response}.
-  - When {sourceStream} completes: complete {responseStream}.
+- Let {responseStream} be a new _event stream_.
+- When {sourceStream} emits {sourceValue}:
+  - Let {response} be the result of running
+    {ExecuteSubscriptionEvent(subscription, schema, variableValues,
+    sourceValue)}.
+  - If internal {error} was raised:
+    - Cancel {sourceStream}.
+    - Complete {responseStream} with {error}.
+  - Otherwise emit {response} on {responseStream}.
+- When {sourceStream} completes normally:
+  - Complete {responseStream} normally.
+- When {sourceStream} completes with {error}:
+  - Complete {responseStream} with {error}.
```

Comment:
Does this wording imply that when the {responseStream} completes with error, it must do so in the same manner as when the {sourceStream} does? We [have a request at graphql-js](https://github.com/graphql/graphql-js/issues/4001) for emitting well formated `GraphQLError`s instead of throwing. I would read the specification permissively, as in the manner in which the {responseStream} completes with said error is up to the implementation, allowing that change.

Once could argue instead/additionally that it belongs within the transport protocol specification.

Tagging @leebyron @enisdenjo @aseemk from the linked issue.

Reply:
From the SSE perspective, and in the [GraphQL over SSE RFC](https://github.com/graphql/graphql-over-http/blob/main/rfcs/GraphQLOverSSE.md), errors are emitted over the stream as an event.

This is because browser native EventSource will not report back to the user non-200 responses in detail - it'll just say "error" - leaving the user in dark. Furthermore, if the error occurs after a while _during_ the stream, the connection is already established and the only way to report an issue is in form of an event.

Same thing applies for WebSockets.

Reply:
I agree that for normal errors in `sourceStream` we should capture them and emit an `{ errors: [ ... ] }` payload rather than terminating with an error.

```suggestion
  - Let {errors} be a list containing {error}.
  - Let {response} be an unordered map containing {errors}.
  - Emit {response} on {responseStream}.
  - Complete {responseStream} normally.
```

However, this is a change in behavior (not an editorial change like the rest of this PR), so perhaps this should be raised _after_ this editorial only change is merged - see https://github.com/graphql/graphql-spec/pull/1127 for a suitable follow-up PR.

Reply:
> I agree that for normal errors in sourceStream we should capture them and emit an { errors: [ ... ] } payload rather than terminating with an error.

Generally agree with using the GraphQL response structure to convey errors but that makes me question why we would need anything else. Is there any specific reason to keep the `Complete {responseStream} with {error}` [just above](https://github.com/graphql/graphql-spec/pull/1099/files#diff-0f02d73330245629f776bb875e5ca2b30978a716732abca136afdd028d5cd33cR314)?

Reply:
I'm not certain, but I imagine that an internal error implies that something fundamental has gone wrong and thus we're no longer sure that we can even complete the request correctly - a bit like returning a "500 Internal Server Error" rather than a 4xx error in HTTP: "something really really went wrong and we're just giving up right here and exiting immediately". Perhaps we can't even be trusted to form a GraphQL response payload at this point - perhaps it was attempting to form such a payload that caused the internal error in the first place?

Reply:
https://github.com/graphql/graphql-spec/pull/1099#discussion_r1643169652

See also above where @leebyron explains….

Reply:
@yaacovCR I've seen that comment but not being super familiar with graphql-js, I'm not sure I understand all the subtleties of it. Once we have https://github.com/graphql/graphql-js/pull/4302, are there still places where graphql-js can throw? I thought this was the last one.

Reply:
Should we move this discussion to #1127 as it does not seem to be editorial but a change in behavior?

Reply:
This is also very related to https://github.com/graphql/graphql-js/issues/4001#issuecomment-2506524425. Happy to consolidate everything to #1127

Reply:
_*Not really making a point, just giving context from the transport perspective._

When dealing with HTTP request-response format, we have 2 places conveying information (as per GQL over HTTP): the body (_data_) and the headers (_metadata_). As @benjie said, 4XX and 5XX would indicate the severity of the errors in _data_.

However, in streaming transports this is different. When dealing with both WS and SSE, browser native implementations (WebSocket and EventSource respectively) give out less information about termination of requests. WebSocket's close event is limited in size, so you cannot send a big error message; and EventSource will tell you absolutely nothing if the server responds with a non-200 status.

Multipart/incremental-delivery/custom-SSE are different because the user does the heavy-lifting (implementation of the fetcher and parser) and there you certainly can use headers for metadata.